redfish-core/lib/environment_metrics.hpp

```diff
@@ -1,18 +1,102 @@
 #pragma once
 
 #include "app.hpp"
+#include "dbus_utility.hpp"
+#include "error_messages.hpp"
 #include "query.hpp"
 #include "registries/privilege_registry.hpp"
+#include "str_utility.hpp"
 #include "utils/chassis_utils.hpp"
+#include "utils/fan_utils.hpp"
 
 #include <boost/url/format.hpp>
 
+#include <array>
+#include <functional>
 #include <memory>
 #include <optional>
 #include <string>
+#include <string_view>
 
 namespace redfish
 {
+inline void
+    updateFanSensorList(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                        const std::string& chassisId,
+                        const std::string& fanSensorPath, double value)
+{
+    std::string fanSensorName =
+        sdbusplus::message::object_path(fanSensorPath).filename();
+    if (fanSensorName.empty())
+    {
+        messages::internalError(asyncResp->res);
+        return;
+    }
+
+    nlohmann::json::object_t item;
+    item["SensorFanArrayExcerpt"]["DataSourceUri"] = boost::urls::format(
+        "/redfish/v1/Chassis/{}/Sensors/{}", chassisId, fanSensorName);
+    item["SensorFanArrayExcerpt"]["DeviceName"] = "Chassis #" + fanSensorName;
+    item["SensorFanArrayExcerpt"]["SpeedRPM"] = value;
+
+    nlohmann::json& fanSensorList =
+        asyncResp->res.jsonValue["FanSpeedsPercent"];
+    fanSensorList.emplace_back(std::move(item));
+    std::sort(fanSensorList.begin(), fanSensorList.end(),
+              [](const nlohmann::json& c1, const nlohmann::json& c2) {
+        return c1["SensorFanArrayExcerpt"]["DataSourceUri"] <
+               c2["SensorFanArrayExcerpt"]["DataSourceUri"];
+    });
+    asyncResp->res.jsonValue["FanSpeedsPercent@odata.count"] =
+        fanSensorList.size();
+}
+
+inline void
+    getFanSensorValue(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                      const std::string& chassisId,
+                      const std::string& serviceName,
+                      const std::string& fanSensorPath)
+{
+    if (serviceName.empty() || fanSensorPath.empty())
+    {
+        messages::internalError(asyncResp->res);
+        return;
+    }
+
+    sdbusplus::asio::getProperty<double>(
+        *crow::connections::systemBus, serviceName, fanSensorPath,
+        "xyz.openbmc_project.Sensor.Value", "Value",
+        [asyncResp, chassisId,
+         fanSensorPath](const boost::system::error_code& ec, double value) {
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR("D-Bus response error for getFanSensorValue {}",
+                             ec.value());
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        updateFanSensorList(asyncResp, chassisId, fanSensorPath, value);
+    });
+}
+
+inline void
+    getFanSpeedsPercent(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
```

Comment:
It looks like - the upstream seems suggesting to restructure the code - "One callback per async action".
I also made a comment to it.

https://gerrit.openbmc.org/c/openbmc/bmcweb/+/57715/comment/8a18bdf9_f737ce8d/